agents/self_evolving_car/self_evolving_car_agent.py

```diff
@@ -0,0 +1,194 @@
+        ball_state = BallState(
+            Physics(velocity=Vector3(0, 0, 0), location=Vector3(0, -1000, 1200), angular_velocity=Vector3(0, 0, 0)))
+        game_state = GameState(ball=ball_state, cars={self.index: car_state})
+        self.set_game_state(game_state)
+
+        # KILL
+        if (my_car.physics.location.z < 100 or my_car.physics.location.z > 1950 or my_car.physics.location.x < -4000
+            or my_car.physics.location.x > 4000 or my_car.physics.location.y > 5000) and self.frame > 50:
+            self.frame = 5000
+
+        return self.controller_state
+
+    def calc_fitness(self):
+        # CALCULATE MINIMUM DISTANCE TO BALL FOR EACH GENOME
+        min_distance_to_ball = 10000000
+        for i in self.distance_to_ball:
+            if i < min_distance_to_ball:
+                min_distance_to_ball = i
+        self.botList[self.brain].fitness = min_distance_to_ball
+        for i in range(len(self.distance_to_ball)):
+            self.distance_to_ball[i] = 100000
+
+        return min_distance_to_ball
+
+    def avg_best_fitness(self):
+        # CALCULATE AVG FITNESS OF 5 FITTEST (IDENTICAL) GENOMES
+        avg = 0
+        for i in range(5, len(self.botList)):
+            avg += self.botList[i].fitness
+        avg /= 5
+        for i in range(5, len(self.botList)):
+            self.botList[i].fitness = avg
+
+    def calc_fittest(self):
+        temp = 1000000
+        count = -1
+        for i in self.botList:
+            count += 1
+            if i.fitness < temp:
+                temp = i.fitness
+                self.fittest.index = count
+        self.fittest.fitness = temp
+        return self.fittest.index
+
+    def reset(self):
+        # RESET TRAINING ATTRIBUTES AFTER EACH GENOME
+        ball_state = BallState(Physics(velocity=Vector3(0, 0, 0), location=Vector3(self.pos, 5000, 3000),
+                                       angular_velocity=Vector3(0, 0, 0)))
+        car_state = CarState(jumped=False, double_jumped=False, boost_amount=33,
+                             physics=Physics(velocity=Vector3(0, 0, 0), rotation=Rotator(45, 90, 0),
+                                             location=Vector3(0.0, -4608, 500), angular_velocity=Vector3(0, 0, 0)))
+        game_info_state = GameInfoState(game_speed=1)
+        game_state = GameState(ball=ball_state, cars={self.index: car_state}, game_info=game_info_state)
+        self.set_game_state(game_state)
+
+    def selection(self):
+        # COPY FITTEST WEIGHTS TO ALL GENOMES
+        path = "./dict.pt"
+        fittest_dict = torch.save(self.botList[self.fittest.index].model.state_dict(),path)
+        self.botList[0].model.load_state_dict(fittest_dict)
+
+    def mutate(self):
+        # MUTATE FIRST 5 GENOMES
+        for i in self.botList:
+           for p in range(0, 4):
+                mutweight = random.randint(0, 1)
+                for param in i.model.parameters():
+                    param.data = torch.rand(param.data.size())
+
```

Comment:
The mutate() code seems to be fully functional, perhaps this would lead to a solution for selection()